sound/soc/sof/ipc4-topology.c

```diff
@@ -1283,9 +1283,6 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
 					   ALH_MULTI_GTW_BASE;
 				ida_free(&alh_group_ida, group_id);
 			}
-
-			/* clear the node ID */
-			copier_data->gtw_cfg.node_id &= ~SOF_IPC4_NODE_INDEX_MASK;
```

Comment:
The node_id was cleared by this commit : fixup! ASoC: SOF: sof-audio: Fix broken early bclk feature for SSP https://github.com/thesofproject/linux/commit/5e9a8563b4e04bd59ed2c472b274dda5e6a6716d. I didn't get it.

The use_count is zero when the node_id is cleared, or it can't be cleared.
Let's talk about our processing
(1) stream open
(2) stream hw_params,   set up fw module here, use_count ++
(3) stream free,  free and unprepare the fw module here,  use_count --

This bug happens with stream a at  step (3),  stream b at step (1),  so we can't depend on use_count.
If we want to use use_count, we need to set up fw module in stream open, at least part of fw. This is not done by both IPC3 & IPC4

Reply:
> The commit message is rather convoluted: there is nothing remotely close to something being 'reserved'.
> 
> you are not clearing a value, but that does not quite explain the problem, and it's weird to leave something initialized unconditionally. 

Our fw module are initialized by topology setting when SOF is loaded. As I remember node_id is initialized by topology in IPC3. @bardliao  do you remember this ?

Reply:
> > The commit message is rather convoluted: there is nothing remotely close to something being 'reserved'.
> > you are not clearing a value, but that does not quite explain the problem, and it's weird to leave something initialized unconditionally.
> 
> Our fw module are initialized by topology setting when SOF is loaded. As I remember node_id is initialized by topology in IPC3. @bardliao do you remember this ?

For IPC3, yes, the dai_index is from topology.

Reply:
Why don't we set the node_id unconditionally before sending a message which needs it?
If the node_id got cleared in between setting it in the message and sending it then we have way bigger problems with race...

Reply:
and few lines up we do: `ida_free(&alh_group_ida, node_id);` so not only we clear the node_id from the IPC message but we might do an ida free as well?

Reply:
ida free is no harm and is necessary for no share BE case.

Reply:
> Why don't we set the node_id unconditionally before sending a message which needs it? If the node_id got cleared in between setting it in the message and sending it then we have way bigger problems with race...

hi @bardliao , please help to make it clear, thanks!

Reply:
> > Why don't we set the node_id unconditionally before sending a message which needs it? If the node_id got cleared in between setting it in the message and sending it then we have way bigger problems with race...
> 
> hi @bardliao , please help to make it clear, thanks!

I think it should be fine if we set the node_id unconditionally before sending a message which needs it. But @ranj063 is the best person to answer this question.

Reply:
> The node_id was cleared by this commit : **fixup! ASoC: SOF: sof-audio: Fix broken early bclk feature for SSP** [5e9a856](https://github.com/thesofproject/linux/commit/5e9a8563b4e04bd59ed2c472b274dda5e6a6716d). I don't get it.
> 
> The use_count is zero when the node_id is cleared, or it can't be cleared. Let's talk about our processing (1) stream open (2) stream hw_params, set up fw module here, use_count ++ (3) stream free, free and unprepare the fw module here, use_count --
> 
> This bug happens with stream A at step (3), stream B at step (1), so we can't depend on use_count.

@RanderWang I think this observation is incorrect. Is Stream A has already finished with hw_free because of which the widget has been freed and unprepared, wouldn't Stream B's hw_params take care of preparing the widget?

Reply:
@ranj063 check this https://github.com/thesofproject/sof/issues/8056#issuecomment-1899870604.   Stream A free widget before Stream B's hw_params happen.  If Stream B's hw_params happen before Stream A free widget, no issue.
If I misunderstand your idea, please help explain it again. Thanks!

Reply:
The node_id is set by SDW hw_params, not by sof_hw_params.

Reply:
@RanderWang so  it looks like we have a race between hw_free and hw_params for the same BE by two separate FE's isnt it? Not clearing the node ID, while it will work, seems like a hack and not an actual fix. Would it make sense to instead protect the be_dai_hw_params and be_dai_hw_free with the dpcm_mutex_lock for the be instead?
@plbossart what do you think?